File: src/branch.swift
import Swiftline
import Foundation

struct Branch {
  let name: String

  func origin() -> String {
    return "origin/\(name)"
  }
}

func getCurrentBranch() -> Branch? {
  let result = runCommand("git symbolic-ref HEAD")
  let matches = result.stdout.matches(forRegex: "heads\\/(.*)")
  if matches.count == 0 {
    return nil
  } else {
    return Branch(name: matches[0])
  }
}

func setCurrentBranch(_ branch: Branch) {
  print("Switching to branch \(branch.name.s.Bold)...")
  fetch()
  detectChanges()
  resetLocal()
  switchBranch(branch)
  detectAhead()
  resetToOrigin()
}

func detectChanges() {
  let status = gitStatus()
  if status.contains("to be committed") ||
      status.contains("for commit:") ||
      status.contains("Untracked files:") {
    uncommitedChanges()
  }
}

func resetLocal() {
  _ = runCommand("git reset --hard")
}

func switchBranch(_ branch: Branch) {
  _ = runCommand("git checkout \(branch.name)")

  if getCurrentBranch()!.name != branch.name {
    _ = runCommand("git checkout -b \(branch.name)")
  }

  _ = runCommand("git branch --set-upstream-to=origin/\(branch.name)")

  if getCurrentBranch()!.name != branch.name {
    print("🤔  Failed to switch branch".f.Red)
    exit(1)
  }
}

func detectAhead() {
  let status = gitStatus()
  if status.contains("can be fast-forwarded.") ||
      status.contains("is ahead of 'origin") ||
      status.contains(" have diverged") {
    branchesDiverged()
  }
}

// On branch master
// Your branch and 'origin/master' have diverged,
// and have 1 and 1 different commit each, respectively.
//   (use "git pull" to merge the remote branch into yours)
// nothing to commit, working directory clean
//
// On branch master
// Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
//   (use "git pull" to update your local branch)
// nothing to commit, working directory clean
//
// On branch master
// Your branch is ahead of 'origin/master' by 1 commit.
//   (use "git push" to publish your local commits)
// nothing to commit, working directory clean
func branchesDiverged() {
  print("\n😱  You appear to have a diverged branch:".f.Red)
  let matches = gitStatus().matches(forRegex: "(Your branch .*\\s*.*)\\s*\\(")
  for match in matches { print(match) }

  promptKeepLocal()
}

func resetToOrigin() {
  let origin = getCurrentBranch()!.origin()
  let reset = runCommand("git reset --hard \(origin)")

  if reset.exitStatus == 0 {
    print("Using remote branch".f.Green)
  } else {
    print("Using local branch (no origin branch found)".f.Green)
  }
}

func printCurrentBranch() {
  let branchName = getCurrentBranch()?.name ?? "no branch"
  print("On branch \(branchName.s.Bold)")
}

func printRecentBranches() {
  let command = runCommand("git for-each-ref --sort=-committerdate --format=\"%(refname)\" --count=30 refs/heads/ refs/remotes")
  let references = command.stdout.components(separatedBy: "\n")

  let commits: [Commit] = references.map {
    Commit(message: "", sha: $0.clearQuotes())
  }

  let strings = commits.map {
    $0.printableFormat("%Cgreen%cr%Creset %C(yellow)%d%Creset %C(bold blue)<%an>%Creset%n")
  }

  var lastStr = ""
  for str in strings {
    if str != lastStr { print(str) } // avoid dups
    lastStr = str
  }
}

func uncommitedChanges() {
  print("\n😱  You appear to have uncommited changes:".f.Red)
  printGitStatus()
  promptContinueAnyway()
}

func gitStatus() -> String {
  return runCommand("git status").stdout
}

func printGitStatus(preceedingNewline: Bool = false) {
  let diff = gitStatus().matches(forRegex: "\t([a-z ]*:.*)")
  if diff.count > 0 && preceedingNewline {
    print("")
  }
  for line in diff {
    print("\t\(line)".f.Green)
  }
}

func promptContinueAnyway() {
  if !agree("\nContinue anyway? Changes will be lost\n>") {
    exit(1)
  }
}

func promptKeepLocal() {
  let choice: String

  if options.preferLocal {
    choice = "local"
  } else if options.preferRemote {
    choice = "remote"
  } else {
    choice = choose("Keep remote or local copy?\n>", type: String.self) { settings in
      settings.addChoice("remote") { "remote" }
      settings.addChoice("local") { "local" }
    }
  }

  if choice != "remote" {
    print("Using local branch (user specified)")
    exit(0)
  }
}

func addAll() {
  _ = runCommand("git reset --mixed")
  _ = runCommand("git add . -A")
}

func fetch() {
  let fetch = runCommand("git fetch")
  let error = fetch.stderr
  if (error as NSString).contains("No remote repository specified") {
    print("\n⚠️  No remote repository is setup\n".f.Yellow)
    return
  }
  if fetch.exitStatus != 0 {
    print("\n🤔  Failed to fetch. How's your internet connection?".f.Red)
    exit(1)
  }
}

File: src/branch.rs
use crate::git;
use crate::git_helpers;
use crate::status;
use crate::prompt;
use crate::output::{output_line_in_red, output_line_in_green, output_line_in_yellow};
use crate::output::format_as_bold;
use shell::ShellResultExt;

pub fn set_branch(target_branch: &str) {
    println!("Switching to branch {}...", format_as_bold(target_branch));

    git::fetch();
    if git_helpers::has_uncommitted_files() {
        output_line_in_yellow("\nThere are uncommitted changes");
        git_helpers::print_uncommited_files();
        let user_choice = prompt::continue_anyway();

        match user_choice {
            prompt::Answer::Abort => {
              output_line_in_red("Aborted (user specified)");
              std::process::exit(1);
            },
            prompt::Answer::ContinueAndDiscard => {
              output_line_in_green("Changes discarded (user specified)");
              let _ = git::reset(git::ResetMode::Hard, "");
              switch_branch(target_branch);
              detect_ahead(target_branch);
              reset_to_origin(target_branch);
            }
            prompt::Answer::ContinueAndCarry => {
              output_line_in_green("Changes being carried over (user specified)");
              let commit = stash_as_a_commit();

              switch_branch(target_branch);
              detect_ahead(target_branch);
              reset_to_origin(target_branch);

              let _ = git::cherry_pick(&commit);
              let _ = git::reset(git::ResetMode::Soft, "HEAD^");
            }
        }
    }

    let _ = git_helpers::add_all();
}

fn stash_as_a_commit() -> String {
  let _ = git_helpers::add_all();
  let _ = git::commit("TEMP_BRANCH_COMMIT");
  let temp_commit_hash = git::log(&"-1 --pretty=format:\"%H\"").trim().to_string();
  let _ = git::reset(git::ResetMode::Hard, "HEAD~1");

  temp_commit_hash
}

fn switch_branch(target_branch: &str) {
    let _ = git::checkout(&target_branch);
    let current_branch = git_helpers::current_branch();

    if current_branch != target_branch {
        let _ = git::checkout(&format!("-b {}", target_branch));
    }

    let _ = git::branch(&format!("--set-upstream-to=origin/{}", target_branch));

    // refetch the current branch to confirm whether the switch was successful
    let current_branch = git_helpers::current_branch();

    if current_branch != target_branch {
        output_line_in_red("Failed to switch branch");
        std::process::exit(1);
    }
}

fn detect_ahead(target_branch: &str) {
    let status = git::status(&"");

    // On branch master
    // Your branch and 'origin/master' have diverged,
    // and have 1 and 1 different commit each, respectively.
    //   (use "git pull" to merge the remote branch into yours)
    // nothing to commit, working directory clean
    //
    // On branch master
    // Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
    //   (use "git pull" to update your local branch)
    // nothing to commit, working directory clean
    //
    // On branch master
    // Your branch is ahead of 'origin/master' by 1 commit.
    //   (use "git push" to publish your local commits)
    // nothing to commit, working directory clean
    let diverged = status.contains("can be fast-forwarded.") ||
        status.contains("is ahead of 'origin") ||
        status.contains(" have diverged");

    if diverged {
        output_line_in_yellow("\nLocal and remote branches have diverged");
        status::print_commits_behind_and_ahead_of_origin(&target_branch);
        prompt::keep_local();
    }
}

fn reset_to_origin(target_branch: &str) {
    let origin = format!("origin/{}", target_branch);
    let output = git::reset(git::ResetMode::Hard, &origin);

    if output.code() == 0 {
        output_line_in_green("Using remote branch");
    } else {
        output_line_in_green("Using local branch (no origin branch found)");
    }
}
